misk-crypto/src/main/kotlin/misk/crypto/FieldLevelEncryptionPacket.kt

```diff
@@ -0,0 +1,358 @@
+package misk.crypto
+
+import com.google.common.io.ByteStreams
+import misk.crypto.FieldLevelEncryptionPacket.EntryType
+import java.io.ByteArrayInputStream
+import java.io.ByteArrayOutputStream
+import java.io.DataInputStream
+import java.io.DataOutput
+import org.apache.commons.collections4.map.CaseInsensitiveMap
+
+/**
+ * [FieldLevelEncryptionPacket] defines a packet format that encodes a ciphertext and associated metadata, typically
+ * additionally authenticated data (AAD). Its designed to be extendable and compact.  Additionally authenticated data is
+ * a context consisting of key: value pairs. Keys are canonicalized to be lower-case.
+ *
+ * The format is a variation on the basic Type-Length-Value encoding scheme.
+ *
+ * The packet is defined as
+ *
+ * [2 byte context bitmask][ ... 1 or more TLV packets ... ]
+ *
+ * Where almost every TLV packet is:
+ *
+ *   [1 byte type][2 byte length][N byte value]
+ *
+ * with the exception of type [EntryType.PAYLOAD_REMAINDER], in which case the size is skipped and payload is considered
+ * to be until the end of the packet.
+ *
+ */
+class FieldLevelEncryptionPacket private constructor(
```

Comment:
Is this a custom encoding scheme? Is there an alternative in tink? would protos work?